controller/destination/server.go

```diff
@@ -152,3 +227,33 @@ func toAddrSet(endpoints []common.TcpAddress) *pb.AddrSet {
 	}
 	return &pb.AddrSet{Addrs: addrs}
 }
+
+func splitDNSName(dnsName string) ([]string, error) {
+	// If the name is fully qualified, strip off the final dot.
+	if strings.HasSuffix(dnsName, ".") {
+		dnsName = dnsName[:len(dnsName)-1]
+	}
+
+	labels := strings.Split(dnsName, ".")
+
+	// Rejects any empty labels, which is especially important to do for
+	// the beginning and the end because we do matching based on labels'
+	// relative positions.
+	for _, l := range labels {
+		if l == "" {
+			return []string{}, errors.New("Empty label in DNS name: " + dnsName)
+		}
+	}
+	return labels, nil
+}
+
+func maybeStripSuffixLabels(input []string, suffix []string) ([]string, bool) {
+	if len(input) <= 1+len(suffix) {
+		return input, false
+	}
+	n := len(input) - len(suffix)
+	if !reflect.DeepEqual(input[n:], suffix) {
```

Comment:
I chose to do it this way because (1) It is less error-prone (no loop conditions or anything to get wrong), (2) it is shorter, (3) when I went looking for the idiomatic way, this was suggested repeatedly. I think we should keep it as-is.